Support for HTTP Auth on request and Fix for bulk index 'percolate'

mehtryx opened the pull request:
I've been working with your client and ran into two items I felt could be helpful.

1) Current implementation of elasticsearch does not support _percolate during a bulk index, only 'percolate'  I've modfied the code for this exception, should it be fixed this will not be required but will also not break.

2) Many people have asked and look for ways to authenticate...when they do the trick is then having the client support it.  I reviewed the lib your useing for request and it supports http basic auth as well as https protocol.  I've modified the code to allow additional options to be specified which you can use to accomplish requests with http auth inplain text or http auth encrypted or by default connect without authentication at all.

Please review the changes, I've run the vows tests (online and offline) to vet the code as is, further testing in my environment which uses nginx as a reverse proxy with authentication and the use of percolate on my bulk indexes is working great.

Thanks.

pmcorp-keithbenedict committed:
Update for percolate in bulk index operations, added authentication support for Basic HTTP auth and https protocol.

rgrove commented:
Thanks for the patch. Can you please add tests for your changes? I'll make additional code comments inline.

rgrove commented on lib/client.js:
```diff
@@ -41,13 +46,19 @@ function Client(host, options) {
+    this.options = options || {};
 
     this.host    = host || '127.0.0.1';
-    this.options = options || {};
```
Why did `this.options` need to move?

mehtryx replied:
This did need to move, just trying to stay with the DRY principal, the hostname needs to be able to include the auth credentials if used, so rather than accessing options twice... I moved it to before the setting of host.

rgrove replied:
This doesn't change anything though. In either place, this.options still isn't used until later. What am I missing?

mehtryx replied:
err, my bad.

No it did not change anything.  Sorry in my mind it was moved to not repeat...but specifically for these lines:

```
this.host    = host || '127.0.0.1';
// Include auth information if present
if (this.options.auth) {
    this.host = this.options.auth + '@' + this.host ;
}
```

The idea was to keep the changes to this.host together, made more sense than splitting it apart with the line this.options = options || {} between them.

If you want it moved back, I can alter that, I just tied it together with the lines modifing this.host to keep it blocked logically together since there is no reason this.options needed to come after this.host.

So thats the rationale ( just had to rewire my brain to explain what I meant ;) )  Let me know if you want it changed.

rgrove commented on lib/client.js:
```diff
@@ -41,13 +46,19 @@ function Client(host, options) {
     this.host    = host || '127.0.0.1';
-    this.options = options || {};
+    // Include auth informatioh if present
+    if (this.options.auth)
```
Opening brace on the same line, please. Also, typo in comment (s/informatioh/information/).

rgrove commented on lib/client.js:
```diff
@@ -41,13 +46,19 @@ function Client(host, options) {
+    // Include auth informatioh if present
+    if (this.options.auth)
+    {
+        this.host = this.options.auth + '@' + this.host;
```
Do `this.options.auth` and `this.host` need to be encoded here? I don't know off the top of my head.

mehtryx replied:
They do not, the request library takes care of this with the url.parse() method.

rgrove commented on lib/client.js:
```diff
@@ -41,13 +46,19 @@ function Client(host, options) {
     this.options.port || (this.options.port = 9200);
     this.options.timeout || (this.options.timeout = 60000);
-
+    (this.options.protocol == 'https:') || (this.options.protocol = 'http:');
```
An if statement would probably be clearer here, since you're doing a comparison rather than just testing truthiness.

rgrove commented on lib/client.js:
```diff
@@ -72,11 +83,13 @@ Client.prototype = {
     /**
     Base URL for this client, of the form "http://host:port".
 
+	+ Modified to support protocol option for https and http
```
No need to mention this in the API docs.

rgrove commented on lib/client.js:
```diff
@@ -72,11 +83,13 @@ Client.prototype = {
-    get baseUrl() {
-        return 'http://' + this.host + ':' + this.options.port;
+    get baseUrl() {	
+		return this.options.protocol + '//' + this.host + ':' + this.options.port;
```
Reindent please.

rgrove commented on lib/index.js:
```diff
@@ -100,7 +105,8 @@ adhere to one of the following formats.
 @param {Client} client Client instance.
 @param {Object[]} operations Array of operations to perform. See above for a
-    description of the expected object format.
+    description of the expected object format. Note currentlty percolate action 
```
No need to mention this. That's probably something we should leave up to ElasticSearch's docs.

pmcorp-keithbenedict committed:
Updated comments, structure and tests for pull request - percolate on bulk and http auth

mehtryx commented:
I've updated the items mentioned above:
- tabs converted to spaces where it messed up indent
- Recommended code styling changes (Open brace same line to be consistent with rest of code)
- If statement to cover the protocol check
- test cases added, resulting in updates to the setup.sh, tests in both offline and online, all passed.  Please review and see if you have any problems with how and where I updated or added tests.  I've tried to match your structure and style there testing the functionality impacted by the proposed code changes.

pmcorp-keithbenedict committed:
Fixed typo in comment for index.js

rgrove commented:
Pulled, with some cleanup and a few minor changes. Thanks!
